pkg/logql/downstream.go

```diff
@@ -382,43 +381,49 @@ func (ev *DownstreamEvaluator) NewIterator(
 	}
 }
 
+type concatStepEvaluator struct {
+	evaluators []StepEvaluator
+}
+
 // NewConcatStepEvaluator joins multiple StepEvaluators.
 // Contract: They must be of identical start, end, and step values.
-func NewConcatStepEvaluator(evaluators []StepEvaluator) (StepEvaluator, error) {
-	return newStepEvaluator(
-		func() (ok bool, ts int64, vec promql.Vector) {
-			var cur promql.Vector
-			for _, eval := range evaluators {
-				ok, ts, cur = eval.Next()
-				vec = append(vec, cur...)
-			}
-			return ok, ts, vec
-		},
-		func() (lastErr error) {
-			for _, eval := range evaluators {
-				if err := eval.Close(); err != nil {
-					lastErr = err
-				}
-			}
-			return lastErr
-		},
-		func() error {
-			var errs []error
-			for _, eval := range evaluators {
-				if err := eval.Error(); err != nil {
-					errs = append(errs, err)
-				}
-			}
-			switch len(errs) {
-			case 0:
-				return nil
-			case 1:
-				return errs[0]
-			default:
-				return util.MultiError(errs)
-			}
-		},
-	)
+func NewConcatStepEvaluator(evaluators []StepEvaluator) StepEvaluator {
```

Comment:
Can we remove the [preemtive interface](https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a) by returning the concrete implementation?

```suggestion
func NewConcatStepEvaluator(evaluators []StepEvaluator) *concatStepEvaluator {
```

`concatStepEvaluator` implicitly implements `StepEvaluator` already.

Ref: https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Same goes with other implementations.

(`concatStepEvaluator` might need to become `ConcactStepEvaluator`, because an exported function should only return exported objects)

Reply:
Done.